Control if invalid configuration values are swallowed or thrown

SteveDunn opened the pull request:
Implements #73915

SteveDunn committed:
First pass

SteveDunn commented on src/libraries/Microsoft.Extensions.Configuration.Binder/tests/ConfigurationBinderTests.cs:
```diff
@@ -505,6 +505,114 @@ public enum TestSettingsEnum
             Option2,
         }
 
+        public class WhenBindingCollectionsWeCanControlExceptionsWithTheErrorOnUnknownConfigurationProperty
```
I know this doesn't fit in with the existing naming convention, but I was struggling to find suitably terse named to describe the scenario **and** the outcome.  Suggestions very welcome.

SteveDunn commented on src/libraries/Microsoft.Extensions.Configuration.Binder/src/Resources/Strings.resx:
```diff
@@ -132,6 +132,9 @@
     <value>Failed to create instance of type '{0}'.</value>
   </data>
+  <data name="Error_GeneralErrorWhenBinding" xml:space="preserve">
+    <value>'{0}' was set and binding has failed due. The likely cause is an invalid configuration value.</value>
```
```suggestion
    <value>'{0}' was set and binding has failed. The likely cause is an invalid configuration value.</value>
```

SteveDunn committed:
Update src/libraries/Microsoft.Extensions.Configuration.Binder/src/Resources/Strings.resx

tarekgh commented:
@SteveDunn could you please resolve the conflict and address the minor comment on the change? 

Also are you interested to fill the breaking change template as described in the comment https://github.com/dotnet/runtime/pull/77708#issuecomment-1298691802? it is ok if not, I can do it later when we merge.

SteveDunn commented:
Unfortunately I'm not free until the weekend, but if that's ok then I'd be happy to.

SteveDunn committed:
Rename test

SteveDunn commented:
> Added `needs-breaking-change-doc-created` label because this PR has the `breaking-change` label.
> 
> When you commit this breaking change:
> 
> 1. [ ]  Create and link to this PR and the issue a matching issue in the dotnet/docs repo using the [breaking change documentation template](https://github.com/dotnet/docs/issues/new?assignees=gewarren&labels=breaking-change%2CPri1%2Cdoc-idea&template=breaking-change.yml&title=%5BBreaking+change%5D%3A+), then remove this `needs-breaking-change-doc-created` label.
> 2. [ ]  Ask a committer to mail the `.NET Breaking Change Notification` DL.
> 
> Tagging @dotnet/compat for awareness of the breaking change.

* https://github.com/dotnet/docs/issues/32175
* @tarekgh - could you mail  `.NET Breaking Change Notification`?

SteveDunn commented:
> > Added `needs-breaking-change-doc-created` label because this PR has the `breaking-change` label.
> > When you commit this breaking change:
> > 
> > 1. [x]   Create and link to this PR and the issue a matching issue in the dotnet/docs repo using the [breaking change documentation template](https://github.com/dotnet/docs/issues/new?assignees=gewarren&labels=breaking-change%2CPri1%2Cdoc-idea&template=breaking-change.yml&title=%5BBreaking+change%5D%3A+), then remove this `needs-breaking-change-doc-created` label.
> > 2. [x]   Ask a committer to mail the `.NET Breaking Change Notification` DL.
> > 
> > Tagging @dotnet/compat for awareness of the breaking change.
> 
> * [[Breaking change]: ConfigurationBinder raises exceptions if ErrorOnUnknownConfiguration is set docs#32175](https://github.com/dotnet/docs/issues/32175)
> * @tarekgh - could you mail  `.NET Breaking Change Notification`?

I spotted an issue with the grammar - highlighted below:

![image](https://user-images.githubusercontent.com/263416/199841729-72498a70-e871-4a2b-b873-116c43a386d5.png)

SteveDunn commented:
> > Unfortunately I'm not free until the weekend, but if that's ok then I'd be happy to.
> 
> Can at least resolve the merge conflict?

Rebased to fix the conflict and added the breaking change docs.

tarekgh commented:
@dotnet/compat could you please help in fixing the grammar issue mentioned in the comment https://github.com/dotnet/runtime/pull/77708#issuecomment-1302718493?
